### DISCUSSION TITLE
Custom binary operators

### DISCUSSION THREAD

**Discussion by baszalmstra** [SeaQL/sea-query#483] [Ideas]
I would like to use the operators from Postgres' https://www.postgresql.org/docs/current/pgtrgm.html extension. Is it possible to add support for custom (binary) operators? 

I did try to use the following:

```rust
builder.prepare_column_ref(&some_column_ref, &mut sql);
write!(sql, "<->");
let (sign, numbered) = builder.placeholder();
sql.push_param(sign, numbered);
Expr::cust_with_values(&sql.result(), [Value::from(matching)])
```

but later in the pipeline it looks like the expression is retokenized which causes the `<->` operator to become `< ->` (note the extra space). I could also just use the `similarity` function provided by the extension but functions don't use an index which is important for large tables.

I see a few solutions:
* Add a `BinOp::Custom(str)`
* Add the similarity operators to `BinOp` (only for postgres)
* Fix the "retokenization"

WDYT?

**Comment by ikrivosheev** [comment]
@baszalmstra hello! Thank you for the idea) would you like to create PR?

**Reply by baszalmstra** [reply]
Ofcourse, but what do you feel is the best approach?

**Reply by ikrivosheev** [reply]
I think both solutions are good. 

1. Add a  `BinOp::Custom(str)`
2. Add the similarity operators to BinOp (only for postgres)

What are the problems with "retokenization"?

**Reply by baszalmstra** [reply]
Added a PR here to add the operators: https://github.com/SeaQL/sea-query/pull/486

The retokenization seems to happen when you use `BinOper::CustWithValues`. It seems the `<->` operator is tokenized as `<` and `->` which causes the query to be malformed.

**Reply by baszalmstra** [reply]
Actually scratch that, that only seems to be the case for logging the query?
